vertical-pod-autoscaler/initializer/initializer.go

```
@@ -0,0 +1,347 @@
+	pod := obj.(*v1.Pod)
+
+	if !shouldInitialize(pod) {
+		glog.V(4).Infof("Not updating pod: %v", pod.Name)
+		return
+	}
+	initializedPod, err := i.getInitializedPod(pod)
+
+	if err == nil {
+		// TODO(bskiba): Right now this update always fails if we try to change resource requests
+		// as there are checks in place that prevent from changing requests of a created pod.
+		result, err := i.client.CoreV1().Pods(pod.Namespace).Update(initializedPod)
+		if err == nil {
+			glog.V(3).Infof("Successfully updated pod %+v", result)
+			return
+		}
+	}
+	failedPod, err := getFailedPod(pod)
+	if err == nil {
+		i.client.CoreV1().Pods(pod.Namespace).Update(failedPod)
+	} else {
+		log.Fatalf("Unable to mark failed initialization for pod %v: %v", pod.Name, err)
+	}
+
+}
+
+func getFailedPod(pod *v1.Pod) (*v1.Pod, error) {
+	failedPodCopy, err := api.Scheme.Copy(pod)
+	if err != nil {
+		return nil, err
+	}
+	failedPod := failedPodCopy.(*v1.Pod)
+	markInitializationFailure(failedPod)
+	return failedPod, nil
+}
+
+// getInitializedPod returns a pod in an initialized state, updating resource
+// requests if applicable.
+func (i *initializer) getInitializedPod(pod *v1.Pod) (*v1.Pod, error) {
+
+	glog.V(4).Infof("Updating requirements for pod: %v", pod.Name)
+
+	updatedPodCopy, err := api.Scheme.Copy(pod)
+	if err != nil {
+		return nil, err
+	}
+	updatedPod := updatedPodCopy.(*v1.Pod)
+	glog.Infof("UPDATED POD: %+v", updatedPod)
+	markInitializationSuccess(updatedPod)
+
+	vpaConfig := i.getMatchingVPA(pod)
+	if vpaConfig == nil {
+		glog.V(4).Infof("No matching vpa found.")
+		return updatedPod, nil
+	}
+
+	recommendation, err := i.recommender.Get(&pod.Spec)
+	if err != nil || recommendation == nil {
+		if vpaConfig.Status.Recommendation != nil {
+			// fallback to recommendation cached in VPA config
+			recommendation = vpaConfig.Status.Recommendation
+		} else {
+			// no recommendation to apply
+			return updatedPod, nil
```

what if there is an error? we eat the error in this case.